File: test/clients/go/jobs_test.go
// Load / Extract / Copy job coverage through the Go bigquery client.
//
// The session-scoped GCS stub from main_test.go backs every URI; the
// emulator was spawned with STORAGE_EMULATOR_HOST pointed at it, so
// `gs://...` source/destination URIs round-trip through the stub.
//
// Covers:
//   - BL-083 Load CSV with autodetect + explicit schema
//   - BL-084 Load NDJSON with autodetect
//   - BL-085 Load Parquet (round-trip via extract → load)
//   - BL-094 Extract to CSV + NDJSON
//   - BL-095 Copy table (default + WRITE_TRUNCATE)
package bqlocal_test

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"sort"
	"strings"
	"testing"

	"cloud.google.com/go/bigquery"
)

func randSuffix(t *testing.T) string {
	t.Helper()
	b := make([]byte, 4)
	if _, err := rand.Read(b); err != nil {
		t.Fatalf("rand: %v", err)
	}
	return hex.EncodeToString(b)
}

// ---------------------------------------------------------------------------
// Load: CSV
// ---------------------------------------------------------------------------

func TestLoadCSVFromGCSWithAutodetect(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	bucket := "go-load-" + randSuffix(t)
	gcsStub.put(bucket, "orders.csv", []byte(
		"order_id,customer,amount,delivered\n"+
			"1,Alice,9.99,true\n"+
			"2,Bob,12.50,false\n"+
			"3,Charlie,7.00,true\n"), "text/csv")

	src := bigquery.NewGCSReference("gs://" + bucket + "/orders.csv")
	src.SourceFormat = bigquery.CSV
	src.AutoDetect = true
	loader := client.Dataset("ds").Table("orders").LoaderFrom(src)
	job, err := loader.Run(ctx)
	if err != nil {
		t.Fatalf("loader.Run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("load wait: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT count(*)::INT64 AS n FROM `ds.orders`"))
	if err != nil {
		t.Fatalf("count query: %v", err)
	}
	if rows[0][0].(int64) != 3 {
		t.Errorf("loaded rows = %d, want 3", rows[0][0])
	}
}

func TestLoadCSVWithExplicitSchema(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	bucket := "go-load-" + randSuffix(t)
	gcsStub.put(bucket, "notes.csv", []byte("id,note\n1,first\n2,second\n"), "text/csv")

	src := bigquery.NewGCSReference("gs://" + bucket + "/notes.csv")
	src.SourceFormat = bigquery.CSV
	src.SkipLeadingRows = 1
	src.Schema = bigquery.Schema{
		{Name: "id", Type: bigquery.StringFieldType, Required: true},
		{Name: "note", Type: bigquery.StringFieldType},
	}
	job, err := client.Dataset("ds").Table("notes").LoaderFrom(src).Run(ctx)
	if err != nil {
		t.Fatalf("load run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("load wait: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT id, note FROM `ds.notes` ORDER BY id"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	want := [][]string{{"1", "first"}, {"2", "second"}}
	for i, w := range want {
		if rows[i][0].(string) != w[0] || rows[i][1].(string) != w[1] {
			t.Errorf("row[%d] = %v, want %v", i, rows[i], w)
		}
	}
}

// ---------------------------------------------------------------------------
// Load: NDJSON
// ---------------------------------------------------------------------------

func TestLoadNDJSONFromGCSWithAutodetect(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	bucket := "go-load-" + randSuffix(t)
	gcsStub.put(bucket, "events.ndjson",
		[]byte(`{"id":1,"kind":"click"}`+"\n"+`{"id":2,"kind":"view"}`+"\n"),
		"application/x-ndjson")

	src := bigquery.NewGCSReference("gs://" + bucket + "/events.ndjson")
	src.SourceFormat = bigquery.JSON
	src.AutoDetect = true
	job, err := client.Dataset("ds").Table("events").LoaderFrom(src).Run(ctx)
	if err != nil {
		t.Fatalf("load run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("load wait: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT id, kind FROM `ds.events` ORDER BY id"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if rows[0][0].(int64) != 1 || rows[0][1].(string) != "click" {
		t.Errorf("row 0 = %v, want [1 click]", rows[0])
	}
	if rows[1][0].(int64) != 2 || rows[1][1].(string) != "view" {
		t.Errorf("row 1 = %v, want [2 view]", rows[1])
	}
}

// ---------------------------------------------------------------------------
// Load: Parquet (round-trip via extract → load)
// ---------------------------------------------------------------------------

func TestLoadParquetFromGCS(t *testing.T) {
	ctx := context.Background()
	client, project := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	if err := client.Dataset(project + "_src").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("src dataset: %v", err)
	}
	src := client.Dataset(project + "_src").Table("fixture")
	if err := src.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "label", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create src: %v", err)
	}
	if err := src.Inserter().Put(ctx, []*struct {
		ID    int64  `bigquery:"id"`
		Label string `bigquery:"label"`
	}{{ID: 1, Label: "alpha"}, {ID: 2, Label: "beta"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}

	bucket := "go-load-" + randSuffix(t)
	dest := bigquery.NewGCSReference("gs://" + bucket + "/fixture.parquet")
	dest.DestinationFormat = bigquery.Parquet
	ej, err := src.ExtractorTo(dest).Run(ctx)
	if err != nil {
		t.Fatalf("extract run: %v", err)
	}
	if _, err := ej.Wait(ctx); err != nil {
		t.Fatalf("extract wait: %v", err)
	}

	loadSrc := bigquery.NewGCSReference("gs://" + bucket + "/fixture.parquet")
	loadSrc.SourceFormat = bigquery.Parquet
	loadSrc.AutoDetect = true
	lj, err := client.Dataset("ds").Table("from_parquet").LoaderFrom(loadSrc).Run(ctx)
	if err != nil {
		t.Fatalf("load run: %v", err)
	}
	if _, err := lj.Wait(ctx); err != nil {
		t.Fatalf("load wait: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT id, label FROM `ds.from_parquet` ORDER BY id"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	want := [][]any{{int64(1), "alpha"}, {int64(2), "beta"}}
	for i, w := range want {
		if rows[i][0] != w[0] || rows[i][1] != w[1] {
			t.Errorf("row[%d] = %v, want %v", i, rows[i], w)
		}
	}
}

// ---------------------------------------------------------------------------
// Extract
// ---------------------------------------------------------------------------

func TestExtractCSVWritesHeaderAndRows(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	tbl := client.Dataset("ds").Table("users")
	if err := tbl.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "name", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	if err := tbl.Inserter().Put(ctx, []*struct {
		ID   int64  `bigquery:"id"`
		Name string `bigquery:"name"`
	}{{ID: 1, Name: "Alice"}, {ID: 2, Name: "Bob"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	bucket := "go-extract-" + randSuffix(t)
	dest := bigquery.NewGCSReference("gs://" + bucket + "/users.csv")
	dest.DestinationFormat = bigquery.CSV
	job, err := tbl.ExtractorTo(dest).Run(ctx)
	if err != nil {
		t.Fatalf("extract run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("extract wait: %v", err)
	}
	stored, ok := gcsStub.get(bucket, "users.csv")
	if !ok {
		t.Fatal("extracted object missing from stub")
	}
	lines := []string{}
	for _, line := range strings.Split(string(stored.bytes), "\n") {
		if line != "" {
			lines = append(lines, line)
		}
	}
	if lines[0] != "id,name" {
		t.Errorf("header = %q, want id,name", lines[0])
	}
	rest := lines[1:]
	sort.Strings(rest)
	if rest[0] != "1,Alice" || rest[1] != "2,Bob" {
		t.Errorf("body = %v, want [1,Alice 2,Bob]", rest)
	}
}

func TestExtractNDJSONRoundTrip(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	tbl := client.Dataset("ds").Table("events")
	if err := tbl.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "kind", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create: %v", err)
	}
	if err := tbl.Inserter().Put(ctx, []*struct {
		ID   int64  `bigquery:"id"`
		Kind string `bigquery:"kind"`
	}{{ID: 7, Kind: "click"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	bucket := "go-extract-" + randSuffix(t)
	dest := bigquery.NewGCSReference("gs://" + bucket + "/events.ndjson")
	dest.DestinationFormat = bigquery.JSON
	job, err := tbl.ExtractorTo(dest).Run(ctx)
	if err != nil {
		t.Fatalf("extract run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("extract wait: %v", err)
	}
	stored, ok := gcsStub.get(bucket, "events.ndjson")
	if !ok {
		t.Fatal("extract output missing")
	}
	var parsed []map[string]any
	for _, line := range strings.Split(string(stored.bytes), "\n") {
		if line == "" {
			continue
		}
		var m map[string]any
		if err := json.Unmarshal([]byte(line), &m); err != nil {
			t.Fatalf("json decode: %v (line=%q)", err, line)
		}
		parsed = append(parsed, m)
	}
	if len(parsed) != 1 {
		t.Fatalf("parsed = %d lines, want 1", len(parsed))
	}
	// BQ NDJSON extract emits INT64 as a string (matches real BQ output).
	if parsed[0]["id"] != "7" || parsed[0]["kind"] != "click" {
		t.Errorf("ndjson row = %v, want {id:7, kind:click}", parsed[0])
	}
}

// ---------------------------------------------------------------------------
// Copy
// ---------------------------------------------------------------------------

func TestCopyTableRoundTrip(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	src := client.Dataset("ds").Table("source")
	dst := client.Dataset("ds").Table("dest")
	if err := src.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "label", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create src: %v", err)
	}
	if err := src.Inserter().Put(ctx, []*struct {
		ID    int64  `bigquery:"id"`
		Label string `bigquery:"label"`
	}{{ID: 1, Label: "one"}, {ID: 2, Label: "two"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	job, err := dst.CopierFrom(src).Run(ctx)
	if err != nil {
		t.Fatalf("copy run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("copy wait: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT id, label FROM `ds.dest` ORDER BY id"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	if rows[0][0].(int64) != 1 || rows[0][1].(string) != "one" {
		t.Errorf("row 0 = %v, want [1 one]", rows[0])
	}
	if rows[1][0].(int64) != 2 || rows[1][1].(string) != "two" {
		t.Errorf("row 1 = %v, want [2 two]", rows[1])
	}
}

func TestCopyWithWriteTruncateOverwritesDestination(t *testing.T) {
	ctx := context.Background()
	client, _ := newClient(t)
	if err := client.Dataset("ds").Create(ctx, &bigquery.DatasetMetadata{}); err != nil {
		t.Fatalf("dataset: %v", err)
	}
	src := client.Dataset("ds").Table("src")
	dst := client.Dataset("ds").Table("dst")
	for _, ref := range []*bigquery.Table{src, dst} {
		if err := ref.Create(ctx, &bigquery.TableMetadata{
			Schema: bigquery.Schema{{Name: "v", Type: bigquery.IntegerFieldType}},
		}); err != nil {
			t.Fatalf("create %s: %v", ref.TableID, err)
		}
	}
	type vrow struct {
		V int64 `bigquery:"v"`
	}
	if err := src.Inserter().Put(ctx, []*vrow{{V: 1}, {V: 2}, {V: 3}}); err != nil {
		t.Fatalf("insert src: %v", err)
	}
	if err := dst.Inserter().Put(ctx, []*vrow{{V: 999}}); err != nil {
		t.Fatalf("insert dst: %v", err)
	}
	copier := dst.CopierFrom(src)
	copier.WriteDisposition = bigquery.WriteTruncate
	job, err := copier.Run(ctx)
	if err != nil {
		t.Fatalf("copy run: %v", err)
	}
	if _, err := job.Wait(ctx); err != nil {
		t.Fatalf("copy wait: %v", err)
	}
	rows, err := readAll(t, client.Query("SELECT v FROM `ds.dst` ORDER BY v"))
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	want := []int64{1, 2, 3}
	for i, w := range want {
		if got := rows[i][0].(int64); got != w {
			t.Errorf("row[%d] = %d, want %d", i, got, w)
		}
	}
}

File: test/clients/java/src/test/java/local/bigquery/JobsTest.java
package local.bigquery;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.CopyJobConfiguration;
import com.google.cloud.bigquery.CsvOptions;
import com.google.cloud.bigquery.DatasetInfo;
import com.google.cloud.bigquery.ExtractJobConfiguration;
import com.google.cloud.bigquery.Field;
import com.google.cloud.bigquery.FieldValueList;
import com.google.cloud.bigquery.FormatOptions;
import com.google.cloud.bigquery.InsertAllRequest;
import com.google.cloud.bigquery.JobInfo;
import com.google.cloud.bigquery.LoadJobConfiguration;
import com.google.cloud.bigquery.Schema;
import com.google.cloud.bigquery.StandardSQLTypeName;
import com.google.cloud.bigquery.StandardTableDefinition;
import com.google.cloud.bigquery.TableId;
import com.google.cloud.bigquery.TableInfo;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.junit.jupiter.api.Test;

/** Load / extract / copy jobs through the GCS stub — mirrors jobs_test.go. */
class JobsTest {
  private static String bucket(String prefix) {
    return prefix + "-" + UUID.randomUUID().toString().substring(0, 8);
  }

  private static void createTable(BigQuery bq, String table, Field... fields) {
    bq.create(TableInfo.of(TableId.of("ds", table), StandardTableDefinition.of(Schema.of(fields))));
  }

  @Test
  void loadCsvFromGcsWithAutodetect() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    String b = bucket("java-load");
    Emu.gcsPut(
        b,
        "orders.csv",
        ("order_id,customer,amount,delivered\n1,Alice,9.99,true\n2,Bob,12.50,false\n3,Charlie,7.00,true\n")
            .getBytes(StandardCharsets.UTF_8),
        "text/csv");

    bq.create(
            JobInfo.of(
                LoadJobConfiguration.newBuilder(
                        TableId.of("ds", "orders"), "gs://" + b + "/orders.csv", FormatOptions.csv())
                    .setAutodetect(true)
                    .build()))
        .waitFor();

    List<FieldValueList> rows = Emu.rows(bq, "SELECT count(*) AS n FROM `ds.orders`");
    assertEquals(3L, rows.get(0).get("n").getLongValue());
  }

  @Test
  void loadCsvWithExplicitSchema() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    String b = bucket("java-load");
    Emu.gcsPut(b, "notes.csv", "id,note\n1,first\n2,second\n".getBytes(StandardCharsets.UTF_8), "text/csv");

    Schema schema =
        Schema.of(
            Field.newBuilder("id", StandardSQLTypeName.STRING).setMode(Field.Mode.REQUIRED).build(),
            Field.of("note", StandardSQLTypeName.STRING));
    bq.create(
            JobInfo.of(
                LoadJobConfiguration.newBuilder(
                        TableId.of("ds", "notes"),
                        List.of("gs://" + b + "/notes.csv"),
                        CsvOptions.newBuilder().setSkipLeadingRows(1L).build())
                    .setSchema(schema)
                    .build()))
        .waitFor();

    List<FieldValueList> rows = Emu.rows(bq, "SELECT id, note FROM `ds.notes` ORDER BY id");
    assertEquals("1", rows.get(0).get("id").getStringValue());
    assertEquals("first", rows.get(0).get("note").getStringValue());
    assertEquals("2", rows.get(1).get("id").getStringValue());
    assertEquals("second", rows.get(1).get("note").getStringValue());
  }

  @Test
  void loadNdjsonFromGcsWithAutodetect() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    String b = bucket("java-load");
    Emu.gcsPut(
        b,
        "events.ndjson",
        "{\"id\":1,\"kind\":\"click\"}\n{\"id\":2,\"kind\":\"view\"}\n".getBytes(StandardCharsets.UTF_8),
        "application/x-ndjson");

    bq.create(
            JobInfo.of(
                LoadJobConfiguration.newBuilder(
                        TableId.of("ds", "events"),
                        "gs://" + b + "/events.ndjson",
                        FormatOptions.json())
                    .setAutodetect(true)
                    .build()))
        .waitFor();

    List<FieldValueList> rows = Emu.rows(bq, "SELECT id, kind FROM `ds.events` ORDER BY id");
    assertEquals(1L, rows.get(0).get("id").getLongValue());
    assertEquals("click", rows.get(0).get("kind").getStringValue());
    assertEquals(2L, rows.get(1).get("id").getLongValue());
    assertEquals("view", rows.get(1).get("kind").getStringValue());
  }

  @Test
  void loadParquetFromGcsRoundTrip() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    createTable(
        bq,
        "fixture",
        Field.of("id", StandardSQLTypeName.INT64),
        Field.of("label", StandardSQLTypeName.STRING));
    bq.insertAll(
        InsertAllRequest.newBuilder(TableId.of("ds", "fixture"))
            .addRow(Map.of("id", 1, "label", "alpha"))
            .addRow(Map.of("id", 2, "label", "beta"))
            .build());

    String b = bucket("java-load");
    bq.create(
            JobInfo.of(
                ExtractJobConfiguration.newBuilder(
                        TableId.of("ds", "fixture"), "gs://" + b + "/fixture.parquet")
                    .setFormat("PARQUET")
                    .build()))
        .waitFor();
    bq.create(
            JobInfo.of(
                LoadJobConfiguration.newBuilder(
                        TableId.of("ds", "from_parquet"),
                        "gs://" + b + "/fixture.parquet",
                        FormatOptions.parquet())
                    .setAutodetect(true)
                    .build()))
        .waitFor();

    List<FieldValueList> rows = Emu.rows(bq, "SELECT id, label FROM `ds.from_parquet` ORDER BY id");
    assertEquals(1L, rows.get(0).get("id").getLongValue());
    assertEquals("alpha", rows.get(0).get("label").getStringValue());
    assertEquals(2L, rows.get(1).get("id").getLongValue());
    assertEquals("beta", rows.get(1).get("label").getStringValue());
  }

  @Test
  void extractCsvWritesHeaderAndRows() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    createTable(
        bq,
        "users",
        Field.of("id", StandardSQLTypeName.INT64),
        Field.of("name", StandardSQLTypeName.STRING));
    bq.insertAll(
        InsertAllRequest.newBuilder(TableId.of("ds", "users"))
            .addRow(Map.of("id", 1, "name", "Alice"))
            .addRow(Map.of("id", 2, "name", "Bob"))
            .build());

    String b = bucket("java-extract");
    bq.create(
            JobInfo.of(
                ExtractJobConfiguration.newBuilder(
                        TableId.of("ds", "users"), "gs://" + b + "/users.csv")
                    .setFormat("CSV")
                    .build()))
        .waitFor();

    byte[] out = Emu.gcsGet(b, "users.csv");
    assertNotNull(out, "extracted object missing");
    List<String> lines = new ArrayList<>();
    for (String line : new String(out, StandardCharsets.UTF_8).split("\n")) {
      if (!line.isEmpty()) {
        lines.add(line);
      }
    }
    assertEquals("id,name", lines.get(0));
    List<String> rest = lines.subList(1, lines.size());
    rest.sort(null);
    assertEquals("1,Alice", rest.get(0));
    assertEquals("2,Bob", rest.get(1));
  }

  @Test
  void extractNdjsonRoundTrip() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    createTable(
        bq,
        "events",
        Field.of("id", StandardSQLTypeName.INT64),
        Field.of("kind", StandardSQLTypeName.STRING));
    bq.insertAll(
        InsertAllRequest.newBuilder(TableId.of("ds", "events"))
            .addRow(Map.of("id", 7, "kind", "click"))
            .build());

    String b = bucket("java-extract");
    bq.create(
            JobInfo.of(
                ExtractJobConfiguration.newBuilder(
                        TableId.of("ds", "events"), "gs://" + b + "/events.ndjson")
                    .setFormat("NEWLINE_DELIMITED_JSON")
                    .build()))
        .waitFor();

    byte[] out = Emu.gcsGet(b, "events.ndjson");
    assertNotNull(out, "extract output missing");
    List<String> lines =
        Arrays.stream(new String(out, StandardCharsets.UTF_8).split("\n"))
            .filter(s -> !s.isEmpty())
            .toList();
    assertEquals(1, lines.size());
    // BQ NDJSON extract emits INT64 as a string.
    assertTrue(lines.get(0).contains("\"id\":\"7\""), () -> "line: " + lines.get(0));
    assertTrue(lines.get(0).contains("\"kind\":\"click\""), () -> "line: " + lines.get(0));
  }

  @Test
  void copyTableRoundTrip() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    createTable(
        bq,
        "source",
        Field.of("id", StandardSQLTypeName.INT64),
        Field.of("label", StandardSQLTypeName.STRING));
    bq.insertAll(
        InsertAllRequest.newBuilder(TableId.of("ds", "source"))
            .addRow(Map.of("id", 1, "label", "one"))
            .addRow(Map.of("id", 2, "label", "two"))
            .build());

    bq.create(
            JobInfo.of(
                CopyJobConfiguration.of(TableId.of("ds", "dest"), TableId.of("ds", "source"))))
        .waitFor();

    List<FieldValueList> rows = Emu.rows(bq, "SELECT id, label FROM `ds.dest` ORDER BY id");
    assertEquals("one", rows.get(0).get("label").getStringValue());
    assertEquals("two", rows.get(1).get("label").getStringValue());
  }

  @Test
  void copyWithWriteTruncateOverwritesDestination() throws Exception {
    BigQuery bq = Emu.client(Emu.uniqueProject());
    bq.create(DatasetInfo.newBuilder("ds").build());
    createTable(bq, "src", Field.of("v", StandardSQLTypeName.INT64));
    createTable(bq, "dst", Field.of("v", StandardSQLTypeName.INT64));
    bq.insertAll(
        InsertAllRequest.newBuilder(TableId.of("ds", "src"))
            .addRow(Map.of("v", 1))
            .addRow(Map.of("v", 2))
            .addRow(Map.of("v", 3))
            .build());
    bq.insertAll(
        InsertAllRequest.newBuilder(TableId.of("ds", "dst")).addRow(Map.of("v", 999)).build());

    bq.create(
            JobInfo.of(
                CopyJobConfiguration.newBuilder(TableId.of("ds", "dst"), TableId.of("ds", "src"))
                    .setWriteDisposition(JobInfo.WriteDisposition.WRITE_TRUNCATE)
                    .build()))
        .waitFor();

    List<FieldValueList> rows = Emu.rows(bq, "SELECT v FROM `ds.dst` ORDER BY v");
    assertEquals(3, rows.size());
    assertEquals(1L, rows.get(0).get("v").getLongValue());
    assertEquals(3L, rows.get(2).get("v").getLongValue());
  }
}
